helm/v1beta1/types_helm_chart_repository.go

```diff
@@ -72,17 +71,45 @@ type ConnectionConfig struct {
 	// It is used as a trust anchor to validate the TLS certificate presented by the remote server.
 	// The key "ca-bundle.crt" is used to locate the data.
 	// If empty, the default system roots are used.
-	// The namespace for this config map is openshift-config.
+	// The namespace for this config map is openshift-config for Cluster scope repositoty if not specified by the user.
+	// For ProjectHelmChartRepository the default will be the namespace where the repository is getting instantiated.
```

Comment:
They do have different name for cluster scope and namespace scope

Reply:
The point here is, that because the spec is shared between `HelmChartRepository` and `ProjectHelmChartRepository`, you are adding fields to the cluster scoped version and the namespace scoped version simultaneously. So now, you are adding the ability for a use in project foo, using a `ProjectHelmChartRepository`, to request a secret from project bar, thus creating a privilege escalation.

You should have a separate set of Golang structures for the cluster and namespace scoped CRDs so that you don't create additional fields one on of the two, that don't make sense, but do make sense on the other.

I'm not sure where that change would lie in the API compatibility guidelines for OpenShift though

Are you intending for users to be able to use this field for both namespaced and non namespaced helm charts?

Reply:
yeah i'm struggling to understand the intent.

is it to allow the clusterscoped chart repositories to get their CA from a NS other than openshift-config?

or is it to allow namespace scoped chart repositories to get their CA from a NS other than the local NS?

https://issues.redhat.com/browse/HELM-343 seems to imply the intent is to allow the NS-scoped repositories to get their CA from another NS.  Why is getting it from the local NS (instead of from openshift-config) insufficient?

Either way, assuming the intent is only to enable namespace specifying for one of the two, then you shouldn't be sharing the field definitions between the two types, as David and Joel both noted, and the ability to ref a secret in another NS creates privilege escalation concerns (also as noted by both of them).

Reply:
I agree we i will  need to add two  different structs for connection config for Project Helm Chart Repository and Helm Chart Repository. The helm chart repository ca/tls data will always be searched in openshift-config namespace. Need to revert that change.
For Project Helm Chart Repository we will like to give the user the experience to declare the namespace where the ca/tls data are created . By doing so we are infact providing  enhanced security to the secrets/cm and he can control who has access to the namespace where these resources are located . While interacting with tls supported PHCR , if the user has specified namespace where the ca and tls are located then we search for those resources in that namespace else we look in the current namespace passed by the ui.

Reply:
>  if the user has specified namespace where the ca and tls are located then we search for those resources in that namespace 

How are you going to ensure the user in question has permission to view secrets in that namespace?

What happens if the user loses those permissions?

@deads2k do we have any other examples of apis that allow this sort of "reach across"?  I can envision justifications where it could be "safe", as long as the intermediary controller is the thing consuming the secret content directly and not exposing the content to the user who referenced it (though at a minimum it likely exposes the existence of a secret w/ that name in that namespace)

The only example i can think of offhand is using node credentials to pull images.  Which does have some similarities in that it's allowing a user to consume a secret they aren't allowed to actually view, to do things.  (And in this case, it's not even allowing a user to "use" credentials, it's allowing them to try to validate a certificate, so the scenario where a user uses this api to point to a random secret and uses it to do bad things, is hard to envision, because all they can really do is trick the controller into trying to use that random secret to validate/trust a TLS cert).

That said, i'm a strong fan of "start simple" here, and just having projectrepos get their trustbundle from a local secret.

Reply:
I have made changes to project helm chart repository pointing to a different struct referred to as ConnectionConfigNamespaceScoped .
As per your question about `How are you going to ensure the user in question has permission to view secrets in that namespace?`i think that should be the end user responsibility to provide permission to the other users in the namespace otherwise the console backend won't be able to read the secret/cm and return error.

Reply:
By privalage escalation are you trying to convey that like if a user has access to  only namespace X , then he should not be able to create a Project Helm Chart Repository CR with specifying the metadata.namespace as X and spec.connectionconfig.ca.namespace as Y (to which he does not even have access)

Reply:
That becomes difficult as you have to perform a subject access review on every create/update. I think the general recommendation (and existing pattern within Kube) is that namespaced objects, shouldn't have namespaced references, instead should just be able to access secrets/configmaps within their own namespace

Reply:
> That becomes difficult as you have to perform a subject access review on every create/update. 

and not only the create/update of the ChartRepository, but also if the RBAC changes in such a way that the user no longer has access to the secret in namespace Y then you need to take some action to stop their usage of the secret (which in turn could unexpectedly break people if the person who changed the access doesn't even realize the secret was being referenced from elsewhere)

Reply:
> If they do not have the required permission they will not be able to perform any CRUD operation on the secret/cm and hence no escalation or the handling of scenarios when permission is removed.

this is only true if all access to the referenced secret is done using the users credentials, and not an intermediary component/controller that has its own credentials which may have more privileges than the user.

So far it's unclear (to me) what consumes this chartrepository object and follows/uses the secret ref and how we ensure it does so only using the users creds (or the creds of an SA the user has permission to use).  And of course even introducing an api like this exposes the risk that someone else implements an escalating controller, even if your design does not.

APIs like this can also expose knowledge that a particular namespace or secret exists, even if the user can't get access to it, which is something we generally try to avoid (e.g. the user tries to ref secret X in namespace Foo and get told they don't have permission to access the secret vs the secret doesn't exist.  This can be mitigated w/ appropriate error/denial messages, though)